src/ray/gcs/gcs_client/gcs_client.cc

```diff
@@ -260,90 +266,155 @@ bool GcsClient::GetGcsServerAddressFromRedis(redisContext *context,
+        callback();
+      }
+    });
     break;
   case rpc::GcsServiceFailureType::GCS_SERVER_RESTART:
     // If GCS sever address has changed, reconnect to GCS server and redo
     // subscription.
-    ReconnectGcsServer();
-    // If using GCS server for pubsub, resubscribe to GCS publishers.
-    resubscribe_func_(RayConfig::instance().gcs_grpc_based_pubsub());
-    // Resend resource usage after reconnected, needed by resource view in GCS.
-    node_resource_accessor_->AsyncReReportResourceUsage();
+    result = ReconnectGcsServerAsync([this, callback]() {
+      RAY_LOG(INFO) << "GCS_SERVER_RESTART callback is called.";
+      // If using GCS server for pubsub, resubscribe to GCS publishers.
+      resubscribe_func_(RayConfig::instance().gcs_grpc_based_pubsub());
+      // Resend resource usage after reconnected, needed by resource view in GCS.
+      node_resource_accessor_->AsyncReReportResourceUsage();
+      if (callback) {
+        callback();
+      }
+    });
     break;
   default:
     RAY_LOG(FATAL) << "Unsupported failure type: " << type;
     break;
   }
+
+  return result;
 }
 
-void GcsClient::ReconnectGcsServer() {
+bool GcsClient::ReconnectGcsServerAsync(const std::function<void()> callback) {
+  if (disconnected_) {
+    return false;
+  }
+  {
+    absl::MutexLock l(&reconnect_callbacks_mutex_);
+    pending_reconnect_callbacks_.emplace_back(callback);
+  }
+  if (reconnect_in_progress_) {
+    RAY_LOG(INFO) << "GcsClient reconnect in progress, the callback is pended.";
+    return false;
+  }
+  reconnect_in_progress_ = true;
+  DoReconnect(absl::Now());
+  return true;
+}
+
+void GcsClient::DoReconnect(absl::Time start) {
+  if (disconnected_) {
+    ExecutePendingCallbacks();
+    reconnect_in_progress_ = false;
+    return;
+  }
   std::pair<std::string, int> address;
-  auto timeout_s =
-      absl::Seconds(RayConfig::instance().gcs_rpc_server_reconnect_timeout_s());
-  auto start = absl::Now();
-  auto reconnected = false;
-  while (absl::Now() - start < timeout_s) {
-    if (disconnected_) {
+  if (get_server_address_func_(&address)) {
+    // After GCS is restarted, the gcs client will reestablish the connection. At
+    // present, every failed RPC request will trigger `ReconnectGcsServerAsync`. In order
+    // to avoid repeated connections in a short period of time, we add a protection
+    // mechanism: if the address does not change (meaning gcs server doesn't restart),
+    // the connection can be made at most once in
+    // `minimum_gcs_reconnect_interval_milliseconds` milliseconds.
+    if (last_reconnect_address_ == address &&
+        (current_sys_time_ms() - last_reconnect_timestamp_ms_) <
+            RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()) {
+      RAY_LOG(DEBUG)
+          << "Repeated reconnection in "
+          << RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()
+          << " milliseconds, return directly.";
+      ExecutePendingCallbacks();
```

Comment:
In the original code, subsequent operations will be performed regardless of whether the reconnection is successful or not. See: https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_client/gcs_client.cc#L284
So we should keep the same logic, invoke callbacks when reconnection is not successful, otherwise, some test cases fails.